MdeModulePkg/Universal/Disk/DiskIoDxe/DiskIo.c

```diff
@@ -226,6 +226,15 @@ DiskIoDriverBindingStart (
            This->DriverBindingHandle,
            ControllerHandle
            );
+
+    if (gDiskIoPrivateDataTemplate.BlockIo2 != NULL) {

```

Comment:
It looks like there are scenarios where gDiskIoPrivateDataTemplate.BlockIo2 could be non NULL from a different Disk I/O Start call that fails and the current Disk I/O Start that fails would close an unrelated Block I/O 2 protocol.  Using this global variable in the OpenProtocol() calls at the beginning of this function is not a good idea.  The template global should be treated as a const and perhaps should be changed to const and instead add some local variables for the pointers to the Block I/O Protocol and Block I/O 2 Protocol and assign them to the allocated version after AllocateCopyPool().